M42: Prevent unwanted pin mode switch to output

tpruvot opened the pull request:
Add several basic checks and fixes to M42 command (DIRECT_PIN_CONTROL)

Also, M42 S1 was never set on STM32 arch, if the led pin is a normal digital output
Value was reset to 0 on analogWrite call...

if we set M0/M2 or other input types, ignore S parameter (write)

For the documentation, add two pins modes M4 & M5 to handle common types (like input analog & Open Drain output mode on STM32)

It's also made to prevent switching automatically an output of this kind to a normal output

M4 is less important, could be removed, except to just set/reset the mode during tests. Need some ADC/Analog setup etc i think

tpruvot committed:
M42: Prevent unwanted pin mode switch to output

Also, M42 S1 was never set on STM32 arch, if the led pin is a normal digital output
Value was reset to 0 on analogWrite call...

tpruvot committed:
Forgot an input type...

thinkyhead commented:
For compatibility with AVR we might want to use `SET_INPUT_PULLUP` for case 2. I'm not sure if `pinMode` alone is exactly the same.
```cpp
#define SET_INPUT_PULLUP(IO)  do{ _SET_INPUT(IO); _WRITE(IO, HIGH); }while(0)
```

thinkyhead committed:
Update M42.cpp

tpruvot commented:
seems arduino constants... cf https://prusa3d.github.io/Prusa-Firmware-Buddy-Doc/dc/de5/wiring__constants_8h.html

well i just tried to find a non STM32 framework...

tpruvot commented:
anyway, we dont need to keep framework constant values, more Marlin doc about M42 ;) https://marlinfw.org/docs/gcode/M042.html

tpruvot commented:
you want to write on an input ? you sure its not the reverse.. like set to 1 & switch to input mode ?

tpruvot commented:
well, can be done later... i already have many env extends in my fork to handle the different printers... it will work like it is right now ;) just have to explain the -DSERVO_OPEN_DRAIN=0 to the few users who modded their old card for a bltouch (need some unsoldering on ZMAX, so its limited)

tpruvot committed:
M42: handle OD outputs for maple/STM32F1 too

On this framework, arduino modes are enums, and so.. "not defined", this trick works...

tpruvot commented:
hmm, look like this github branch is broken now...

tested that on the F1 HAL :

```
commit cb0b9cc12530c45e78f74247570018947be43868 (HEAD -> M42_stm32f1)
Author: Tanguy Pruvot <tanguy.pruvot@gmail.com>
Date:   Tue Aug 3 20:53:03 2021 +0200

    M42: handle OD outputs for maple/STM32F1 too

diff --git a/Marlin/src/gcode/control/M42.cpp b/Marlin/src/gcode/control/M42.cpp
index ff140f2cf7..4f14bb5556 100644
--- a/Marlin/src/gcode/control/M42.cpp
+++ b/Marlin/src/gcode/control/M42.cpp
@@ -31,6 +31,10 @@
   #include "../../module/temperature.h"
 #endif

+#ifdef MAPLE_STM32F1
+  #include "../../HAL/STM32F1/fastio.h"
+#endif
+
 void protected_pin_err() {
   SERIAL_ERROR_MSG(STR_ERR_PROTECTED_PIN);
 }
@@ -64,11 +68,15 @@ void GcodeSuite::M42() {
       #ifdef INPUT_PULLDOWN
         case 3: pinMode(pin, INPUT_PULLDOWN); avoidWrite = true; break;
       #endif
-      #ifdef INPUT_ANALOG
+      #ifdef INPUT_ANALOG // STM32
         case 4: pinMode(pin, INPUT_ANALOG); avoidWrite = true; break;
+      #elif defined(MAPLE_STM32F1)
+        case 4: _SET_MODE(pin, GPIO_INPUT_ANALOG); avoidWrite = true; break;
       #endif
-      #ifdef OUTPUT_OPEN_DRAIN
+      #ifdef OUTPUT_OPEN_DRAIN // STM32
         case 5: pinMode(pin, OUTPUT_OPEN_DRAIN); break;
+      #elif defined(MAPLE_STM32F1)
+        case 5: _SET_OUTPUT_OD(pin); break;
       #endif
       default: SERIAL_ECHOLNPGM("Invalid Pin Mode"); return;
     }

```

tpruvot commented on Marlin/src/gcode/control/M42.cpp:
```diff
@@ -55,13 +62,20 @@ void GcodeSuite::M42() {
   if (parser.seenval('M')) {
     switch (parser.value_byte()) {
-      case 0: pinMode(pin, INPUT); break;
+      case 0: pinMode(pin, INPUT); avoidWrite = true; break;
```
maybe we could just put "return;" instead of break on the inputs... but will skip the error message...

thinkyhead replied:
I thing the error message is a good thing.

thinkyhead commented:
> (fatal: Could not read from remote repository.)

I get that error if I try to push to a fork of a fork of Marlin. But no problems for a direct fork of Marlin.

tpruvot commented:
yea dunno what happened, fixed now... even browsing this branch on github website was doing error 500

tpruvot commented:
There may be another was for these enums vs define... could be put in our HAL in a .h or... something else ;)

thinkyhead committed:
Tweak and test push

tpruvot commented:
no no, _SET_MODE is not required anymore... cf the last commit pushed here...

thinkyhead commented:
> There may be another solution for these enums vs define…

It makes sense to patch up the cases we know about. Hopefully we will be able to drop Maple support altogether soon, but for the next minor release it will still be included.

tpruvot commented:
https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/cores/maple/stm32f1/wirish_digital_f1.cpp#L58

tpruvot commented:
i meant, we could put these 3 lines in our F1 HAL : 
+  #define INPUT_PULLDOWN INPUT_PULLDOWN
+  #define INPUT_ANALOG INPUT_ANALOG
+  #define OUTPUT_OPEN_DRAIN OUTPUT_OPEN_DRAIN

tpruvot commented:
to note about maple, it will stay stable... unlike STM32 one ;) v2.0 migration will be a hell/per board job

thisiskeithb commented:
> to note about maple, it will stay stable... unlike STM32 one ;) v2.0 migration will be a hell/per board job. But i will be there for longer one

Staying on Maple wasn't doing us any favors.

tpruvot commented:
its not what i said, handling it was mandatory for new F2/F3/F4/F7 ... boards anyway

tpruvot commented on Marlin/src/gcode/control/M42.cpp:
```diff
@@ -68,12 +68,20 @@ void GcodeSuite::M42() {
+          _SET_MODE(pin, GPIO_INPUT_ANALOG);
+        #endif
+        avoidWrite = true;
+        break;
```
imo revert this commit, will skip default: SERIAL_ECHOLNPGM("Invalid Pin Mode");

thinkyhead committed:
Untweak and test push
